### PR TITLE
Fixed #28334 -- Added LRU cache for hstore/citext oids

### PR DESCRIPTION
**Pull request by exslim**
Ticket [28334](https://code.djangoproject.com/ticket/28334)

### CODE REVIEW

**Comment by timgraham** [comment]
Claude marked the ticket as "Needs tests". Do you think it's feasible to test this? Perhaps with `assertNumQueries()`?

**Comment by exslim** [comment]
assertNumQueries is a good idea. I'm already working on it. Also, does it make any sense to test that OIDs cache will be cleared after CREATE/DROP EXTENSION? To do this test, I'll have to manually send CREATE/DROP EXTENSION via cursor. Will this break tests that rely on hstore and run in parallel?

**Comment by exslim** [comment]
@timgraham - Unfortunately, assertNumQueries can't properly log queries made from HstoreAdapter since psycopg internally it uses native cursor [1] that can't be wrapped by django. It's hard to make this thing testable.

[1] https://github.com/psycopg/psycopg2/blob/master/lib/extras.py#L770-L772

**Comment by timgraham** [comment]
I would be okay committing this optimization without a test. @claudep, did you have any specific concerns or ideas about what you wanted tested when you updated the ticket?

**Comment by exslim** [comment]
For what is worth, citext oid part is totally testable. I'll add tests for this.

File: django/contrib/postgres/signals.py [hunk 1]
```diff
@@ -1,14 +1,43 @@
...
+
+
+@functools.lru_cache()
+def get_hstore_oids(connection_alias):
+    with connections[connection_alias].cursor() as cursor:
+        # Retrieve oids of hstore and hstore array.
+        cursor.execute("SELECT t.oid, typarray "
+                       "FROM pg_type t "
+                       "JOIN pg_namespace ns ON typnamespace = ns.oid "
+                       "WHERE typname = 'hstore'")
```
Comment (adamchainz) [inline django/contrib/postgres/signals.py]: cleaner as a triple quoted string (`"""`), SQL isn't whitespace sensitive

**Comment by exslim** [comment]
@timgraham - I think the patch is ready for review/checkin. I added tests for both hstore and citext parts.

File: tests/postgres_tests/test_hstore.py [hunk 2]
```diff
@@ -20,6 +21,32 @@ class HStoreTestCase(PostgreSQLTestCase):
...
+    def test_citext_at_most_zero_queries(self):
+        from django.contrib.postgres.signals import get_citext_oids
+        # oids cached after app's ready() call
+        with self.assertNumQueries(0):
+            get_citext_oids(connection.alias)
+
+    def test_hstore_return_values(self):
+        from django.contrib.postgres.signals import get_hstore_oids
+        oids, array_oids = get_hstore_oids(connection.alias)
+        self.assertTrue(isinstance(oids, tuple))
```
Comment (timgraham) [inline tests/postgres_tests/test_hstore.py]: Use `assertIsInstance()`. Is it worth checking something like that it's a tuple integers of a certain length?

File: tests/postgres_tests/test_hstore.py [hunk 3]
```diff
@@ -20,6 +21,32 @@ class HStoreTestCase(PostgreSQLTestCase):
     pass
 
 
+class HstoreOIDTests(HStoreTestCase):
+
+    def test_hstore_at_most_zero_queries(self):
+        from django.contrib.postgres.signals import get_hstore_oids
```
Comment (timgraham) [inline tests/postgres_tests/test_hstore.py]: Are the inner imports needed? Can they go in the try/except block at the top of the file?

File: django/contrib/postgres/signals.py [hunk 4]
```diff
@@ -1,14 +1,43 @@
...
 from psycopg2 import ProgrammingError
 from psycopg2.extras import register_hstore
 
+from django.db import connections
+
+
+@functools.lru_cache()
+def get_hstore_oids(connection_alias):
+    with connections[connection_alias].cursor() as cursor:
+        # Retrieve oids of hstore and hstore array.
```
Comment (timgraham) [inline django/contrib/postgres/signals.py]: Better as a docstring?

File: django/contrib/postgres/signals.py (hunk 1 above)
Comment (timgraham) [inline django/contrib/postgres/signals.py]: I don't have a strong opinion about the quote style but use hanging indent as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

File: django/contrib/postgres/operations.py [hunk 5]
```diff
@@ -15,13 +17,19 @@ def database_forwards(self, app_label, schema_editor, from_state, to_state):
         if schema_editor.connection.vendor != 'postgresql':
             return
         schema_editor.execute("CREATE EXTENSION IF NOT EXISTS %s" % schema_editor.quote_name(self.name))
+        # Clear cached oids since they are not valid anymore
```
Comment (timgraham) [inline django/contrib/postgres/operations.py]: A bit simpler could be "Clear cached, stale oids."

**Comment by timgraham** [comment]
@exslim, can you address the latest comments?

**Comment by exslim** [comment]
@timgraham - will do

File: tests/postgres_tests/test_hstore.py (hunk 2 above)
Comment (exslim) [reply tests/postgres_tests/test_hstore.py]: Agreed, `assertIsInstance ()` is better choice. Checking for certain length makes very little sense.

File: django/contrib/postgres/signals.py (hunk 4 above)
Comment (exslim) [reply django/contrib/postgres/signals.py]: probably not

**Comment by timgraham** [comment]
I pushed the test idea from my earlier comment in a second commit (perhaps it was unclear?). I don't know much about how OIDs work though -- could this test be invalid in some cases?

**Comment by exslim** [comment]
I would test all OIDs are integers, but i wouldn't test the length - citext[] OID query could easily result in empty tuple.

**Comment by timgraham** [comment]
For my own knowledge, could you explain when the citext[] OID query would return an empty tuple?

**Comment by exslim** [comment]
It would certainly return empty tuple if you didn't run `CREATE EXTENSION citext;`

It's not the case for `postgres_tests` as migration creates all necessary extensions (especially `hstore` and `citext`). So yeah, length=1 check shouldn't fail there. Also wondering if query could return two or more OIDs on servers with 3rd party extensions that override native types;

**Comment by eukreign** [comment]
This commit broke running unit tests, specifically this line:

https://github.com/django/django/blob/86a18dc46aae3a4a6410e3f3d864fa0ec4e5b2cd/django/contrib/postgres/signals.py#L13

When running unit tests the `alias` that's passed in is `__no_db__` set here:

https://github.com/django/django/blob/86a18dc46aae3a4a6410e3f3d864fa0ec4e5b2cd/django/db/backends/base/base.py#L589

Full traceback I get:

```
[localhost] local: coverage run -p manage.py test -v 2 systori.apps systori.lib
Creating test database for alias 'default' ('test_systori_test')...
Traceback (most recent call last):
File "/usr/local/lib/python3.5/dist-packages/django/db/utils.py", line 167, in ensure_defaults
conn = self.databases[alias]
KeyError: '__no_db__'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "manage.py", line 8, in <module>
execute_from_command_line(sys.argv)
File "/usr/local/lib/python3.5/dist-packages/django/core/management/__init__.py", line 371, in execute_from_command_line
utility.execute()
File "/usr/local/lib/python3.5/dist-packages/django/core/management/__init__.py", line 365, in execute
self.fetch_command(subcommand).run_from_argv(self.argv)
File "/usr/local/lib/python3.5/dist-packages/django/core/management/commands/test.py", line 26, in run_from_argv
super().run_from_argv(argv)
File "/usr/local/lib/python3.5/dist-packages/django/core/management/base.py", line 288, in run_from_argv
self.execute(*args, **cmd_options)
File "/usr/local/lib/python3.5/dist-packages/django/core/management/base.py", line 335, in execute
output = self.handle(*args, **options)
File "/usr/local/lib/python3.5/dist-packages/django/core/management/commands/test.py", line 59, in handle
failures = test_runner.run_tests(test_labels)
File "/usr/local/lib/python3.5/dist-packages/django/test/runner.py", line 601, in run_tests
old_config = self.setup_databases()
File "/usr/local/lib/python3.5/dist-packages/django/test/runner.py", line 548, in setup_databases
self.parallel, **kwargs
File "/usr/local/lib/python3.5/dist-packages/django/test/utils.py", line 176, in setup_databases
serialize=connection.settings_dict.get('TEST', {}).get('SERIALIZE', True),
File "/usr/local/lib/python3.5/dist-packages/django/db/backends/base/creation.py", line 54, in create_test_db
self._create_test_db(verbosity, autoclobber, keepdb)
File "/usr/local/lib/python3.5/dist-packages/django/db/backends/base/creation.py", line 164, in _create_test_db
with self._nodb_connection.cursor() as cursor:
File "/usr/local/lib/python3.5/dist-packages/django/db/backends/base/creation.py", line 27, in _nodb_connection
return self.connection._nodb_connection
File "/usr/local/lib/python3.5/dist-packages/django/db/backends/postgresql/base.py", line 251, in _nodb_connection
nodb_connection.ensure_connection()
File "/usr/local/lib/python3.5/dist-packages/django/db/backends/base/base.py", line 210, in ensure_connection
self.connect()
File "/usr/local/lib/python3.5/dist-packages/django/db/backends/base/base.py", line 191, in connect
connection_created.send(sender=self.__class__, connection=self)
File "/usr/local/lib/python3.5/dist-packages/django/dispatch/dispatcher.py", line 178, in send
for receiver in self._live_receivers(sender)
File "/usr/local/lib/python3.5/dist-packages/django/dispatch/dispatcher.py", line 178, in <listcomp>
for receiver in self._live_receivers(sender)
File "/usr/local/lib/python3.5/dist-packages/django/contrib/postgres/signals.py", line 41, in register_type_handlers
oids, array_oids = get_hstore_oids(connection.alias)
File "/usr/local/lib/python3.5/dist-packages/django/contrib/postgres/signals.py", line 13, in get_hstore_oids
with connections[connection_alias].cursor() as cursor:
File "/usr/local/lib/python3.5/dist-packages/django/db/utils.py", line 199, in __getitem__
self.ensure_defaults(alias)
File "/usr/local/lib/python3.5/dist-packages/django/db/utils.py", line 169, in ensure_defaults
raise ConnectionDoesNotExist("The connection %s doesn't exist" % alias)
django.db.utils.ConnectionDoesNotExist: The connection __no_db__ doesn't exist

Fatal error: local() encountered an error (return code 1) while executing 'coverage run -p manage.py test -v 2 systori.apps systori.lib'
```

**Comment by timgraham** [comment]
Maybe the fix is something like:
```
if connection_alias == '__no_db__':
    return ()  # or (), () for get_hstore_oids()
```
at the start of `get_hstore_oids()` and `get_citext_oids()`. This at least fixes the crash for me.

**Comment by exslim** [comment]
I thought about this, but will this fix length check discussed previously?

BTW, How to reproduce this exception?

**Comment by timgraham** [comment]
I created a PR: https://github.com/django/django/pull/9105. I can reproduce the crash by running `manage.py test` with a project that has `django.contrib.postgres` in `INSTALLED APPS`.

**Pull request merged**